Wikipedia/Code/SaveButtonsController.swift

```diff
@@ -0,0 +1,63 @@
+import UIKit
+
+@objc(WMFSaveButtonsController) class SaveButtonsController: NSObject {
+    
+    var visibleSaveButtons = [Int: SaveButton]()
+    var visibleArticleKeys = [Int: String]()
+    let dataStore: MWKDataStore
+    let savedPagesFunnel = SavedPagesFunnel()
+    
+    required init(dataStore: MWKDataStore) {
+        self.dataStore = dataStore
+        super.init()
+        NotificationCenter.default.addObserver(self, selector: #selector(articleUpdated(notification:)), name: NSNotification.Name.WMFArticleUpdated, object: nil)
+    }
+    
+    deinit {
+        NotificationCenter.default.removeObserver(self)
+    }
+    
+    @objc(willDisplaySaveButton:forArticle:) func willDisplay(saveButton: SaveButton, for article: WMFArticle) {
```

Comment:
I slightly prefer the `@objc` and `func` on separate lines

```
@objc(willDisplaySaveButton:forArticle:)
func willDisplay(saveButton: SaveButton, for article: WMFArticle)
```

I think I've seen that format elsewhere in the code. What do you think?